components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/utils/AppContextCacheUtil.java

```diff
@@ -0,0 +1,103 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.SynapseException;
+import org.wso2.carbon.appmgt.api.AppManagementException;
+import org.wso2.carbon.appmgt.api.model.WebApp;
+import org.wso2.carbon.appmgt.impl.AppMConstants;
+import org.wso2.carbon.appmgt.impl.dao.AppMDAO;
+
+import javax.cache.Cache;
+import javax.cache.Caching;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Utility class for accessing Application contexts cache.
+ * This is primarily used in mediators/handlers in gateway.
+ *
+ * TODO: Usage of the cache makes the caller thread to delay significantly when cache miss.
+ * We might need better handling of this functionality in order to get a good performance.
+ */
+public class AppContextCacheUtil {
+
+    private static final Log log = LogFactory.getLog(AppContextCacheUtil.class);
+
+    /**
+     * Returns the </t/tenantdomain/context/version> or </context/version> to its URL map.
+     * @return
+     */
+    public static Map<String, String> getTenantContextVersionUrlMap() {
+        Cache cache = getAppContextVersionConfigCache();
+        Map<String, String> contextVersionMap = (HashMap<String, String>) cache
+                .get(AppMConstants.APP_CONTEXT_VERSION_CACHE_KEY);
+        if (contextVersionMap == null) {
+            //Set the app context+version list in cache
+            contextVersionMap = getAppContextWithVersion();
+            getAppContextVersionConfigCache()
+                    .put(AppMConstants.APP_CONTEXT_VERSION_CACHE_KEY, contextVersionMap);
+        }
+        return contextVersionMap;
+    }
+
+    /**
+     * get the cache manager
+     * @return
+     */
+    private static Cache getAppContextVersionConfigCache() {
+        return Caching.getCacheManager(AppMConstants.APP_CONTEXT_VERSION_CACHE_MANAGER)
+                .getCache(AppMConstants.APP_CONTEXT_VERSION_CONFIG_CACHE);
+    }
+
+    /**
+     * retrieving context+version string list from database.
+     * @return context+version list
+     */
+    private static Map<String, String> getAppContextWithVersion() {
+        if (log.isDebugEnabled()) {
+            log.debug("Calling getAppContextWithVersion");
+        }
+        try {
+            APIMgtDBUtil.initialize();
+            AppMDAO dao = new AppMDAO();
+            Map<String, String> contextVersion = new HashMap<String, String>();
+            Iterator listI = dao.getAllWebApps().listIterator();
+
+            while (listI.hasNext()) {
+                WebApp app = (WebApp) listI.next();
+                contextVersion.put((app.getContext().startsWith("/") ?
+                        app.getContext() :
+                        "/" + app.getContext()) +
+                        "/" + app.getId().getVersion(), app.getUrl());
+            }
+            return contextVersion;
+        } catch (AppManagementException e) {
+            throw new SynapseException(
```

Comment:
Why do we concatenate parent exception message to the new exception without passing it as an inner exception? Example: new SynapseException(message, e)